example_queries/get_service_name.rs.ref

```diff
@@ -2,50 +2,74 @@ use rpc_lib::rpc::Rpc;
-use petgraph::Outgoing;
+use petgraph::Incoming;
 use graph_utils::utils;
 use graph_utils::iso::find_mapping_shamir_centralized;
+use serde::{Serialize, Deserialize};
+extern crate serde_json;
 
 pub type CodeletType = fn(&Filter, &Rpc) -> Option<Rpc>;
 
-// user defined functions:
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FerriedData {
+    trace_graph: Graph<(String, IndexMap<String, String>), String>,
+    unassigned_properties: Vec<(String, String, String)>, // entity property value
+}
 
+impl FerriedData {
+    fn default() -> FerriedData {
+        FerriedData {
+            trace_graph: Graph::new(),
+            unassigned_properties: Vec::new(),
+            
+        }
+    }
 
-// This represents a piece of state of the filter
-// it contains information that we get from calling getValue in the real filters
-// TODO:  since we no longer use these to define UDFs, we have no need for it
-// to be anything other than a string.  We should simplify it to simply have
-// an Option<String> for every possible envoy value
+    // take any unassigned properties that apply to nodes in the graph,
+    // and associate them with those nodes
+    fn assign_properties(&mut self) {
+        let mut to_delete = Vec::new();
+        for property in &self.unassigned_properties {
+            let node = utils::get_node_with_id(&self.trace_graph, property.0.clone());
+            if node.is_some() {
+                self.trace_graph.node_weight_mut(node.unwrap()).unwrap().1.insert(property.1.clone(), property.2.clone());
+                to_delete.push(property.clone());
+            }
+        }
+        self.unassigned_properties.retain(|x| !to_delete.contains(&x));
 
-#[derive(Clone, Debug)]
-pub struct State {
-    pub type_of_state: Option<String>,
-    pub string_data: Option<String>,
+    }
 }
 
-impl State {
-    pub fn new() -> State {
-        State {
-            type_of_state: None,
-            string_data: None,
+fn put_ferried_data_in_hdrs(fd: &mut FerriedData, hdr: &mut IndexMap<String,String>) {
+    match serde_json::to_string(fd) {
+        Ok(stored_data_string) => {
+            hdr.insert("ferried_data".to_string(), stored_data_string);
         }
-    }
-
-    pub fn new_with_str(str_data: String) -> State {
-        State {
-            type_of_state: Some(String::from("String")),
-            string_data: Some(str_data),
+        Err(e) => {
+            print!("ERROR:  could not translate stored data to json string: {0}\n", e);
         }
     }
 }
 
+
+// user defined functions:
+
+
+
+// This represents a piece of state of the filter
+// it contains information that we get from calling getValue in the real filters
+// TODO:  since we no longer use these to define UDFs, we have no need for it
+// to be anything other than a string.  We should simplify it to simply have
+// an Option<String> for every possible envoy value
+
 #[derive(Clone, Debug)]
 pub struct Filter {
     pub whoami: Option<String>,
```

Comment:
Why is whoami an option? How can that happen?

Reply:
If the call to envoy's data fails, then we have no information on who the node is.  So this is a way of making sure that the call didn't fail, and we're working with real data.

Reply:
If that call fails the filter should just exit early.